File: apps/backend/src/services/message.ts
import { MessageRepository } from '../repositories/message'
import type { RabbitMQService } from './rabbitmq'
import type { SendPersonalMessageDto, HandleDeliveredDto, HandleReadDto, SendGroupMessageDto } from '../dto/message'

export class MessageService {
  constructor(
    private readonly repo: MessageRepository,
    private readonly rabbitmqService: RabbitMQService,
  ) {}

  sendToUser(senderId: number, receiverId: number, content: string) {
    // ensure chats exist
    return this.repo.save(this.repo.create({ content, sender: { id: senderId }, channel: null }))
  }

  getMessagesBetween(context: any, receiverId: number, clearedAt: Date) {
    const senderId = context.user.id
    return this.repo.getMessagesByUserId(senderId, receiverId, clearedAt)
  }

  getMessagesInChannel(channelId: number) {
    return this.repo.getMessagesByChannelId(channelId)
  }

  async sendPersonalMessage(context: any, dto: SendPersonalMessageDto) {
    const senderId = context.user.id

    // Basic validation
    if (!dto.content || !dto.receiverId || !dto.hash) {
      throw new Error('Invalid message payload')
    }

    // Publish to incoming_messages exchange for worker to process
    await this.rabbitmqService.publishToIncoming('personal.message', {
      type: 'MESSAGE_SEND',
      payload: {
        senderId,
        receiverId: dto.receiverId,
        content: dto.content,
        hash: dto.hash,
      },
    })

    return { success: true }
  }

  async sendGroupMessage(context: any, dto: SendGroupMessageDto) {
    const senderId = context.user.id

    // Basic validation
    if (!dto.content || !dto.groupId || !dto.channelId || !dto.hash) {
      throw new Error('Invalid group message payload')
    }

    // Publish to incoming_messages exchange for worker to process
    await this.rabbitmqService.publishToIncoming('group.message', {
      type: 'MESSAGE_SEND',
      payload: {
        senderId,
        groupId: dto.groupId,
        channelId: dto.channelId,
        content: dto.content,
        hash: dto.hash,
      },
    })

    return { success: true }
  }

  async handleDelivered(context: any, dto: HandleDeliveredDto) {
    const receiverId = context.user.id

    // Basic validation
    if (!dto.messageId || !dto.receiverId || !dto.senderId) {
      throw new Error('Invalid delivered payload')
    }

    // Publish to incoming_messages exchange for worker to process
    await this.rabbitmqService.publishToIncoming('personal.delivered', {
      type: 'STATUS_DELIVERED',
      payload: {
        messageId: dto.messageId,
        receiverId,
        senderId: Number(dto.senderId),
      },
    })

    return { success: true }
  }

  async handleRead(context: any, payloads: HandleReadDto[]) {
    const receiverId = context.user.id

    // Basic validation
    if (!Array.isArray(payloads) || payloads.length === 0) {
      throw new Error('Invalid read payload')
    }

    // Publish to incoming_messages exchange for worker to process
    await this.rabbitmqService.publishToIncoming('personal.read', {
      type: 'STATUS_READ',
      payload: payloads.map(p => ({
        messageId: p.messageId,
        senderId: p.senderId,
        receiverId,
      })),
    })

    return { success: true }
  }
}

File: apps/backend-go/internal/service/message.go
package service

import (
	"context"
	"log"

	"github.com/arpansaha13/messaging-system/apps/backend-go/internal/domain"
	"github.com/arpansaha13/messaging-system/apps/backend-go/internal/repository"
)

// MessageService handles message business logic
type MessageService struct {
	messageRepo          *repository.MessageRepository
	messageRecipientRepo *repository.MessageRecipientRepository
	chatRepo             *repository.ChatRepository
	rabbitmqService      *RabbitMQService
}

// NewMessageService creates a new message service
func NewMessageService(messageRepo *repository.MessageRepository, messageRecipientRepo *repository.MessageRecipientRepository, chatRepo *repository.ChatRepository, rabbitmqService *RabbitMQService) *MessageService {
	return &MessageService{
		messageRepo:          messageRepo,
		messageRecipientRepo: messageRecipientRepo,
		chatRepo:             chatRepo,
		rabbitmqService:      rabbitmqService,
	}
}

// SendPersonalMessage publishes a personal message to RabbitMQ for processing
func (s *MessageService) SendPersonalMessage(ctx context.Context, senderID, receiverID int64, content, hash string) error {
	if !s.rabbitmqService.IsConnected() {
		return &domain.InternalError{Message: "RabbitMQ not connected"}
	}

	payload := PersonalMessagePayload{
		SenderID:   senderID,
		ReceiverID: receiverID,
		Content:    content,
		Hash:       hash,
	}

	message := RabbitMQMessage{
		Type:    "MESSAGE_SEND",
		Payload: payload,
	}

	err := s.rabbitmqService.PublishToIncoming("personal.message", message)
	if err != nil {
		log.Printf("failed to publish personal message: %v", err)
		return err
	}

	return nil
}

// SendGroupMessage publishes a group message to RabbitMQ for processing
func (s *MessageService) SendGroupMessage(ctx context.Context, senderID, groupID, channelID int64, content, hash string) error {
	if !s.rabbitmqService.IsConnected() {
		return &domain.InternalError{Message: "RabbitMQ not connected"}
	}

	payload := GroupMessagePayload{
		SenderID:  senderID,
		GroupID:   groupID,
		ChannelID: channelID,
		Content:   content,
		Hash:      hash,
	}

	message := RabbitMQMessage{
		Type:    "MESSAGE_SEND",
		Payload: payload,
	}

	err := s.rabbitmqService.PublishToIncoming("group.message", message)
	if err != nil {
		log.Printf("failed to publish group message: %v", err)
		return err
	}

	return nil
}

// GetMessages retrieves messages between two users
func (s *MessageService) GetMessages(ctx context.Context, userID, receiverID int64, limit, offset int) ([]*repository.MessageWithStatus, error) {
	messages, err := s.messageRepo.GetMessagesByUserId(ctx, userID, receiverID, nil, limit, offset)
	if err != nil {
		log.Printf("failed to get messages: %v", err)
		return nil, err
	}
	return messages, nil
}

// MarkMessageAsDelivered publishes a delivered status update to RabbitMQ
func (s *MessageService) MarkMessageAsDelivered(ctx context.Context, messageID, receiverID, senderID int64) error {
	if !s.rabbitmqService.IsConnected() {
		return &domain.InternalError{Message: "RabbitMQ not connected"}
	}

	payload := DeliveredPayload{
		MessageID:  messageID,
		ReceiverID: receiverID,
		SenderID:   senderID,
	}

	message := RabbitMQMessage{
		Type:    "STATUS_DELIVERED",
		Payload: payload,
	}

	err := s.rabbitmqService.PublishToIncoming("personal.delivered", message)
	if err != nil {
		log.Printf("failed to publish delivered status: %v", err)
		return err
	}

	return nil
}

// MarkMessageAsRead publishes read status updates to RabbitMQ
func (s *MessageService) MarkMessageAsRead(ctx context.Context, messages []ReadPayload) error {
	if !s.rabbitmqService.IsConnected() {
		return &domain.InternalError{Message: "RabbitMQ not connected"}
	}

	message := RabbitMQMessage{
		Type:    "STATUS_READ",
		Payload: messages,
	}

	err := s.rabbitmqService.PublishToIncoming("personal.read", message)
	if err != nil {
		log.Printf("failed to publish read status: %v", err)
		return err
	}

	return nil
}

// GetChannelMessages retrieves messages for a channel
func (s *MessageService) GetChannelMessages(ctx context.Context, channelID int64, limit, offset int) ([]*domain.Message, error) {
	messages, err := s.messageRepo.GetMessagesByChannelID(ctx, channelID, limit, offset)
	if err != nil {
		log.Printf("failed to get channel messages: %v", err)
		return nil, err
	}
	return messages, nil
}
